### DISCUSSION TITLE
Install Teleport behind HAProxy

### DISCUSSION THREAD

**Discussion by damienbutt** [gravitational/teleport#8063] [Q&A] [answered]
Hi all,

I have installed Teleport on an Ubuntu Server 20.04 LTS. 

Following the docs, I have gotten as far as running the `teleport configure` command to generate a default config.

I run HAProxy on my network that does all my SSL terminations using Let's Encrypt. 

Is it possible to generate a self-signed cert with Teleport, or even use it without a cert? And then proxy through to it using HAProxy and the Let's Encrypt cert? This is how I run all my other internal/external services.

Thanks,

D

**Comment by webvictim** [comment, marked as answer]
If you start Teleport without providing a certificate (either via `https_keypairs` in the config file, or using `acme`), its default behaviour is to generate a self-signed certificate and use that when serving the proxy's web UI. As long as haproxy doesn't attempt to validate the self-signed certificate presented by the Teleport backend, it should work fine.

You can use it without a cert if you really want by using `teleport start --insecure-no-tls`, but I don't recommend this.

**Comment by damienbutt** [comment]
I can't seem to get the web UI up on port 443.

Teleport is running. I have opened the necessary ports. Teleport doesn't seem to be listening on any IPv4 ports. Any ideas?

<img width="2158" alt="Screenshot 2021-08-30 at 12 09 39" src="https://user-images.githubusercontent.com/22627489/131330610-73267935-3440-4382-a876-11fe7686e664.png">

<img width="1128" alt="Screenshot 2021-08-30 at 12 10 18" src="https://user-images.githubusercontent.com/22627489/131330636-6479b883-87aa-4fc6-be2a-facff19b0909.png">

<img width="616" alt="Screenshot 2021-08-30 at 12 10 44" src="https://user-images.githubusercontent.com/22627489/131330703-9f979abf-8ccf-455a-971a-bb97ff0c1bdb.png">

<img width="440" alt="Screenshot 2021-08-30 at 12 12 38" src="https://user-images.githubusercontent.com/22627489/131331104-aa1d86da-3828-4c4a-b884-6ed4e326af45.png">

**Reply by webvictim** [reply]
Set this in your config file and restart Teleport.

```yaml
proxy_service:
  web_listen_addr: 0.0.0.0:443
```

**Reply by webvictim** [reply]
For your use case it sounds more like you want to have haproxy listen on 443 and redirect to the Teleport backend on port 3080.

**Reply by damienbutt** [reply]
Yes, HAProxy does my SSL termination. So having an HTTP backend or self-signed HTTPS backend is fine as HAProxy will take care of that. Although I would prefer to run the backend on HTTPS with a self-signed cert if possible.

HAProxy listens on ports 443 and 80.

I have two internal DNS CNAME entries set up.

`teleport.domain.com` and `*.teleport.domain.com`, which both point to `haproxy.internal.domain.com`.

In HAProxy, I have a front end set up to look for `teleport.domain.com` and proxy that traffic to the IP address of my teleport instance on port 3080. This is currently working. The login page is loading with SSL via HAProxy. I haven't done any more than that yet.

Do I need to proxy traffic from `*.teleport.domain.com` through to anything? 

Again, I would prefer to run with a self-signed cert on the backend but I couldn't seem to get that working.

**Reply by webvictim** [reply]
If you don’t specify `https_keypairs` or `acme`, it should generate and use a self-signed cert.

If you want to use Teleport’s application access, you’ll need to proxy the wildcard domain through so you can use subdomains for applications. If you’re not using that, you don’t need to.

**Comment by damienbutt** [comment]
I have only been able to get this working without using a cert. Unless I did something wrong. 

I think the documentation could be a bit better. It doesn't seem to cover the self-cert/no-cert options...

**Comment by damienbutt** [comment]
Also found that without a cert the service listens on port 3080, which isn't mentioned in the docs. Unless I missed that too.
